flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -54,15 +57,16 @@
  * the buffers for the network data transfer. When new local buffer pools are created, the
  * NetworkBufferPool dynamically redistributes the buffers between the pools.
  */
-public class NetworkBufferPool implements BufferPoolFactory, MemorySegmentProvider {
+public class NetworkBufferPool implements BufferPoolFactory, MemorySegmentProvider, AvailabilityProvider {
 
 	private static final Logger LOG = LoggerFactory.getLogger(NetworkBufferPool.class);
 
 	private final int totalNumberOfMemorySegments;
 
 	private final int memorySegmentSize;
 
-	private final ArrayBlockingQueue<MemorySegment> availableMemorySegments;
+	@GuardedBy("availabilityHelper")
+	private final ArrayDeque<MemorySegment> availableMemorySegments;
```

Comment:
Another option is synchronizing the `availableMemorySegments` directly instead of `availabilityHelper`, then we could remove the `@GuardedBy("availabilityHelper")`